File: Creational/ObjectPool/python/test_object_pool.py
import unittest
from object_pool import ObjectPool, ConcreteReusable

class TestObjectPool(unittest.TestCase):
	"""
	Test suite for the ObjectPool implementation.
	"""

	def test_single_borrow_and_return(self):
		"""
		Test borrowing and returning a single object.
		"""
		pool = ObjectPool[ConcreteReusable]()
		obj = ConcreteReusable("Obj1")
		pool.add_object(obj)

		# Borrow the object
		borrowed = pool.borrow_object()
		self.assertEqual(borrowed, obj)

		# Return the object
		pool.return_object(borrowed)

		# Borrow again and verify reuse
		reused = pool.borrow_object()
		self.assertEqual(borrowed, reused)

	def test_fifo_behavior(self):
		"""
		Test that the pool behaves as FIFO (First In, First Out).
		"""
		pool = ObjectPool[ConcreteReusable]()
		obj1 = ConcreteReusable("Obj1")
		obj2 = ConcreteReusable("Obj2")
		pool.add_object(obj1)
		pool.add_object(obj2)

		# Borrow objects in FIFO order
		borrowed1 = pool.borrow_object()
		borrowed2 = pool.borrow_object()
		self.assertEqual(borrowed1, obj1)
		self.assertEqual(borrowed2, obj2)

		# Return objects
		pool.return_object(borrowed1)
		pool.return_object(borrowed2)

		# Borrow again and verify FIFO behavior
		reused1 = pool.borrow_object()
		reused2 = pool.borrow_object()
		self.assertEqual(reused1, obj1)
		self.assertEqual(reused2, obj2)

	def test_empty_pool_borrow(self):
		"""
		Test that borrowing from an empty pool raises an error.
		"""
		pool = ObjectPool[ConcreteReusable]()
		with self.assertRaises(RuntimeError):
			pool.borrow_object()

	def test_return_null_object(self):
		"""
		Test that returning a None object raises an error.
		"""
		pool = ObjectPool[ConcreteReusable]()
		with self.assertRaises(ValueError):
			pool.return_object(None)

	def test_duplicate_return(self):
		"""
		Test that returning the same object twice raises an error.
		"""
		pool = ObjectPool[ConcreteReusable]()
		obj = ConcreteReusable("Obj1")
		pool.add_object(obj)

		# Borrow and return the object
		borrowed = pool.borrow_object()
		pool.return_object(borrowed)

		# Attempt to return the same object again
		with self.assertRaises(ValueError):
			pool.return_object(borrowed)

	def test_pool_size_consistency(self):
		"""
		Test that the pool size is updated correctly after borrow and return operations.
		"""
		pool = ObjectPool[ConcreteReusable]()
		pool.add_object(ConcreteReusable("Obj1"))
		pool.add_object(ConcreteReusable("Obj2"))
		self.assertEqual(pool.get_size(), 2)

		# Borrow an object
		borrowed = pool.borrow_object()
		self.assertEqual(pool.get_size(), 1)

		# Return the object
		pool.return_object(borrowed)
		self.assertEqual(pool.get_size(), 2)

if __name__ == "__main__":
	unittest.main()

File: Creational/ObjectPool/cpp/test_object_pool.cpp
#include "object_pool.hpp"
#include "concrete_reusable.hpp"
#include <gtest/gtest.h>

/**
 * @brief Test borrowing and returning objects in the object pool.
 *
 * This test validates the correct borrowing and returning behavior of the 
 * ObjectPool class. It ensures that returned objects can be reused by the pool 
 * without duplicating or recreating objects unnecessarily.
 */
TEST(ObjectPoolTest, BorrowAndReturnObject) {
	ObjectPool pool;

	// Add reusable objects to the pool
	pool.addObject(std::make_unique<ConcreteReusable>("Obj1"));
	pool.addObject(std::make_unique<ConcreteReusable>("Obj2"));

	// Borrow an object from the pool
	auto obj1 = pool.borrowObject();
	obj1->use();

	// Return the object back to the pool
	pool.returnObject(std::move(obj1));

	// Borrow another object and verify reuse
	auto obj2 = pool.borrowObject();
	obj2->use();

	// Ensure obj1 and obj2 are different instances
	EXPECT_NE(obj1.get(), obj2.get());
}

/**
 * @brief Test borrowing from an empty object pool.
 *
 * This test ensures that attempting to borrow an object from an empty pool
 * throws a `std::runtime_error`. It verifies that the pool correctly handles
 * cases where no reusable objects are available.
 */
TEST(ObjectPoolTest, BorrowFromEmptyPool) {
	ObjectPool pool;

	// Attempt to borrow from an empty pool and expect an exception
	EXPECT_THROW(pool.borrowObject(), std::runtime_error);
}

/**
 * @brief Test multiple borrow and return operations in the object pool.
 *
 * This test validates the behavior of the ObjectPool class under multiple 
 * consecutive borrow and return operations. It ensures that objects are 
 * properly reset and can be reused in a predictable manner.
 */
TEST(ObjectPoolTest, MultipleBorrowAndReturn) {
	ObjectPool pool;

	// Add multiple objects to the pool
	pool.addObject(std::make_unique<ConcreteReusable>("Obj1"));
	pool.addObject(std::make_unique<ConcreteReusable>("Obj2"));

	// Borrow objects from the pool
	auto obj1 = pool.borrowObject();
	auto obj2 = pool.borrowObject();

	obj1->use();
	obj2->use();

	// Return the objects back to the pool
	pool.returnObject(std::move(obj1));
	pool.returnObject(std::move(obj2));

	// Borrow again to ensure returned objects are reused
	auto obj3 = pool.borrowObject();
	auto obj4 = pool.borrowObject();

	obj3->use();
	obj4->use();

	// Ensure the objects are not the same as the previously borrowed instances
	EXPECT_NE(obj1.get(), obj3.get());
	EXPECT_NE(obj2.get(), obj4.get());
}
